x-pack/plugins/case/server/client/comments/add.ts

```diff
@@ -10,135 +10,297 @@ import { pipe } from 'fp-ts/lib/pipeable';
+      }),
+    ],
   });
 
-  const [comments] = await Promise.all([
-    caseService.getAllCaseComments({
-      client: savedObjectsClient,
-      caseId,
-      options: {
-        fields: [],
-        page: 1,
-        perPage: totalCommentsFindByCases.total,
-      },
+  return updatedCase.encode();
+};
+
+async function getCombinedCase(
+  service: CaseServiceSetup,
+  client: SavedObjectsClientContract,
+  id: string
+): Promise<CommentableCase> {
+  const [casePromise, subCasePromise] = await Promise.allSettled([
+    service.getCase({
+      client,
+      id,
     }),
-    userActionService.postUserActions({
-      client: savedObjectsClient,
-      actions: [
-        buildCommentUserActionItem({
-          action: 'create',
-          actionAt: createdDate,
-          actionBy: { username, full_name, email },
-          caseId: myCase.id,
-          commentId: newComment.id,
-          fields: ['comment'],
-          newValue: JSON.stringify(query),
-        }),
-      ],
+    service.getSubCase({
+      client,
+      id,
     }),
   ]);
 
-  return CaseResponseRt.encode(
-    flattenCaseSavedObject({
-      savedObject: {
-        ...myCase,
-        ...updatedCase,
-        attributes: { ...myCase.attributes, ...updatedCase.attributes },
-        version: updatedCase.version ?? myCase.version,
-        references: myCase.references,
-      },
-      comments: comments.saved_objects,
-    })
+  if (subCasePromise.status === 'fulfilled') {
+    if (subCasePromise.value.references.length > 0) {
+      const caseValue = await service.getCase({
+        client,
+        id: subCasePromise.value.references[0].id,
+      });
+      return new CommentableCase({
+        collection: caseValue,
+        subCase: subCasePromise.value,
+        service,
+        soClient: client,
+      });
+    } else {
+      throw Boom.badRequest('Sub case found without reference to collection');
+    }
+  }
+
+  if (casePromise.status === 'rejected') {
+    throw casePromise.reason;
+  } else {
+    return new CommentableCase({ collection: casePromise.value, service, soClient: client });
+  }
+}
+
+interface AddCommentArgs {
+  caseClient: CaseClientHandler;
+  caseId: string;
+  comment: CommentRequest;
+  savedObjectsClient: SavedObjectsClientContract;
+  caseService: CaseServiceSetup;
+  userActionService: CaseUserActionServiceSetup;
+  user: User;
+}
+
+export const addComment = async ({
+  savedObjectsClient,
+  caseService,
+  userActionService,
+  caseClient,
+  caseId,
+  comment,
+  user,
+}: AddCommentArgs): Promise<CollectionWithSubCaseResponse> => {
+  const query = pipe(
+    CommentRequestRt.decode(comment),
+    fold(throwErrors(Boom.badRequest), identity)
   );
+
+  if (isCommentRequestTypeGenAlert(comment)) {
+    return addGeneratedAlerts({
+      caseId,
+      comment,
+      caseClient,
+      savedObjectsClient,
+      userActionService,
+      caseService,
+    });
+  }
+
+  decodeCommentRequest(comment);
```

Comment:
Comment for another PR: TypeScript thinks that the comment is of type `user` and not `alert | user`.